libp2p-swarm/src/transport.rs

```diff
@@ -664,22 +664,37 @@ impl<'a, T, C> UpgradedNode<T, C>
 			}
 		};
 
+		// Try to negotiate the protocol.
+		// Note that failing to negotiate a protocol will never produce a future with an error.
+		// Instead the `stream` will produce an `Ok(Err(...))`.
+		// `stream` can only produce an `Err` if `listening_stream` produces an `Err`.
 		let stream = listening_stream
-			// Try to negotiate the protocol.
 			.and_then(move |(connection, client_addr)| {
-				let upgrade = upgrade.clone();
-				#[inline]
-				fn iter_map<T>((n, t): (Bytes, T)) -> (Bytes, fn(&Bytes,&Bytes)->bool, T) {
-					(n, <Bytes as PartialEq>::eq, t)
-				}
-				let iter = upgrade.protocol_names().map(iter_map);
-				let negotiated = multistream_select::listener_select_proto(connection, iter);
-				negotiated.map(move |(upgrade_id, conn)| (upgrade_id, conn, upgrade, client_addr))
-					.map_err(|err| IoError::new(IoErrorKind::Other, err))
-			})
-			.and_then(|(upgrade_id, connection, upgrade, client_addr)| {
-				upgrade.upgrade(connection, upgrade_id, Endpoint::Listener)
-					.map(|s| (s, client_addr))
+				// Turn the `Result<impl AsyncRead + AsyncWrite, IoError>` into
+				// a `Result<impl Future<Item = impl AsyncRead + AsyncWrite, Error = IoError>, IoError>`
+				let connection = connection.map(|connection| {
+					let upgrade = upgrade.clone();
+					#[inline]
+					fn iter_map<T>((n, t): (Bytes, T)) -> (Bytes, fn(&Bytes,&Bytes)->bool, T) {
```

Comment:
Why can't this be a closure? Even if you've got to manually hint the types you can do `|n: Bytes, t| -> (_, fn(&_, &_) -> bool, _) (n, <Bytes as PartialEq>::eq, t)`?

Reply:
The `listener_select_proto` function requires the iterator to be clonable. Rust closures are unfortunately not clonable, so doing `.map(<closure>)` would make the whole iterator non-clonable.

Reply:
[You can coerce a closure to a pointer](https://play.rust-lang.org/?gist=d7e606ca16ae08de329b90c2e72c1e55&version=stable)

Reply:
I just made the change